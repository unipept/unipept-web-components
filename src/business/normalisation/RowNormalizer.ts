import Normalizer from "./Normalizer";

/**
 * Normalises a grid on a row-per-row basis. For each row the largest and smallest value is selected after which all
 * other values are normalised with respect to these values.
 *
 * @author Pieter Verschaffelt
 */
export default class RowNormalizer implements Normalizer {
    public normalize(data: number[][]): number[][] {
        if (data.length === 0 || data[0].length === 0) {
            return data;
        }

        const output: number[][] = [];
        for (const row of data) {
            const min = Math.min(...row);
            const max = Math.max(...row);

            const newRow = [];
            for (const item of row) {
                if (max - min !== 0) {
                    newRow.push((item - min) / (max - min));
                } else {
                    newRow.push(0);
                }
            }
            output.push(newRow);
        }
        return output;
    }
}
