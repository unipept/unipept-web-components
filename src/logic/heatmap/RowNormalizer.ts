import { Normalizer } from "./Normalizer";

export default class RowNormalizer implements Normalizer {
    public normalize(data: number[][]): number[][] {
        if (data.length === 0 || data[0].length === 0) {
            return data;
        }

        const output: number[][] = [];
        for (const row of data) {
            const min = Math.min(...row);
            const max = Math.max(...row);

            const newRow = [];
            for (const item of row) {
                if (max - min !== 0) {
                    newRow.push((item - min) / (max - min));
                } else {
                    newRow.push(0);
                }
            }
            output.push(newRow);
        }
        return output;
    }
}
