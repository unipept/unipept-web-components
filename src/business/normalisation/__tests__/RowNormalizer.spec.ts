import RowNormalizer from "@/business/normalisation/RowNormalizer";

describe("RowNormalizer", () => {
    it("normalizes a complete grid correctly", () => {
        const normalizer = new RowNormalizer();
        const data = [
            [4, 4, 2, 1],
            [1, 2, 1, 0],
            [4, 3, 0, 4]
        ];

        const normalizedResult = normalizer.normalize(data);
        expect(normalizedResult).toMatchSnapshot();

        // Every row should contain a 1
        expect(normalizedResult.every(row => row.indexOf(1) !== -1)).toBeTruthy();
        // Every row should contain a 0
        expect(normalizedResult.every(row => row.indexOf(0) !== -1)).toBeTruthy();
    });

    it("normalizes a grid with negative values correctly", () => {
        const normalizer = new RowNormalizer();
        const data = [
            [4, 4, -2, 1],
            [-1, 2, 1, 0],
            [-4, 3, 0, 4]
        ];

        const normalizedResult = normalizer.normalize(data);
        // There should no longer be negative values in the result, since everything is mapped onto the [0, 1] interval.
        expect(normalizedResult).toMatchSnapshot();

        // Every row should contain a 1
        expect(normalizedResult.every(row => row.indexOf(1) !== -1)).toBeTruthy();
        // Every row should contain a 0
        expect(normalizedResult.every(row => row.indexOf(0) !== -1)).toBeTruthy();
    });

    it("does not crash on all-zero grids", () => {
        const normalizer = new RowNormalizer();
        const data = [
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0]
        ];

        const normalizedResult = normalizer.normalize(data);
        // There should no longer be negative values in the result, since everything is mapped onto the [0, 1] interval.
        expect(normalizedResult).toMatchSnapshot();
    });
});
