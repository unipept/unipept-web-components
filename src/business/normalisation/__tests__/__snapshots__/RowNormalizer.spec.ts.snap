// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RowNormalizer does not crash on all-zero grids 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`RowNormalizer normalizes a complete grid correctly 1`] = `
Array [
  Array [
    1,
    1,
    0.3333333333333333,
    0,
  ],
  Array [
    0.5,
    1,
    0.5,
    0,
  ],
  Array [
    1,
    0.75,
    0,
    1,
  ],
]
`;

exports[`RowNormalizer normalizes a grid with negative values correctly 1`] = `
Array [
  Array [
    1,
    1,
    0,
    0.5,
  ],
  Array [
    0,
    1,
    0.6666666666666666,
    0.3333333333333333,
  ],
  Array [
    0,
    0.875,
    0.5,
    1,
  ],
]
`;
